src/org/sosy_lab/common/collect/NaiveOrderStatisticMap.java

```diff
@@ -0,0 +1,162 @@
+/*
+ *  SoSy-Lab Common is a library of useful utilities.
+ *  This file is part of SoSy-Lab Common.
+ *
+ *  Copyright (C) 2007-2017  Dirk Beyer
+ *  All rights reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.sosy_lab.common.collect;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ForwardingNavigableMap;
+import com.google.common.collect.Maps;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+/**
+ * An {@link OrderStatisticMap} with naive implementations of its functions.
+ *
+ * <p>The class wraps a {@link NavigableMap} object and delegates all methods inherited from the
+ * <code>NavigableMap</code> interface to that. For the methods particular to the <code>
+ * OrderStatisticMap</code> interface, it provides naive implementations that do not guarantee any
+ * specific performance.
+ *
+ * @param <K> type of the keys of this map. See the Javadoc of {@link OrderStatisticMap} for
+ *     possible constraints on this type
+ * @param <V> type of the values of this map
+ * @see OrderStatisticMap
+ */
+final class NaiveOrderStatisticMap<K, V> extends ForwardingNavigableMap<K, V>
+    implements OrderStatisticMap<K, V>, Serializable {
+
+  private static final long serialVersionUID = -3542217590830996599L;
+
+  private final NavigableMap<K, V> delegate;
+
+  private NaiveOrderStatisticMap(NavigableMap<K, ? extends V> pNavigableMap) {
+    delegate = new TreeMap<>(pNavigableMap);
+  }
+
+  /** Creates a new empty OrderStatisticMap using natural ordering. */
+  static <K, V> NaiveOrderStatisticMap<K, V> createMap() {
```

Comment:
These non-public factory methods are not strictly necessary, the constructor could be made package-private. But if these factory methods help to reduce code, then keep them.

Reply:
I'm not sure whether it's really *less* code, but it circumvents the existence of
`OrderStatisticMap(Map)` and `OrderStatisticMap(NavigableMap)`, and thus is more concrete and less error-prone.